### ISSUE TITLE
Contact about Advice about adaptive algorithm

### ISSUE THREAD

**Issue by Thorhian** [kreso-t/FreeCAD_Mod_Adaptive_Path-Cpp#2]
Hi Kreso!

It wasn’t until today that I tracked down who actually implemented FreeCAD’s only adaptive path algorithm. First of all, thank you. The path workbench would be a lot worse off without your work. 

For quite a long while I have wished that FreeCAD had better algorithms for roughing 3D models, and not be so… stuck in mostly 2.5D (at least without a lot tweaking and head banging). I really wanted a 3D version of an adaptive/constant engagement algorithm, since I want to mill a lot stuff that have 3D features. Other professional software packages have algorithms for this, but FreeCAD’s adaptive is just 2.5D. This eventually pushed me to get out of my comfort zone last year and I started learning how to render and visualize 3D meshes with OpenGL, and now I have a partial prototype that is just a standalone Python application (but I'm using OpenGL shaders for some of the critical parts). 

Is there a good way to communicate with you if you are willing to give advice on the iterative aspects of the path planning? I'm sorry for posting a random issue on your old repo. Thank you again for your addition to the Path workbench.

**Comment by kreso-t** [comment]
Hi Thorhian,

For this I mostly used clipper lib + relatively basic 2D geometry math. Some descriptions and explanations can be found in the FreeCAD forum topic https://forum.freecadweb.org/viewtopic.php?t=30127. However I have not been working on this for a long time, so it would take some time also for me to get back into the topic.

I was planning extend it to 3D by simply slicing the model (in pyhton code using existing FreeCAD API)  with a plane parallel to XY in small steps across the Z axis (i.e. tool step down) and run this same 2d adaptive for each slice. I think this would be easiest way to get it to work for 3D - by reusing existing 2.5D code.

If you want you can contact me via email kresimir.tusek@gmail.com

**Comment by Thorhian** [comment]
Thanks for the response Kreso. I was reading reading that older thread and I found it to be quite amusing, since I have basically come to some similar conclusions about the iterative nature of the algorithm. Problem is, before I started working on this prototype (https://github.com/Thorhian/Path_Tinkering/tree/initial-path-generation), I didn't really have the confidence to delve into FreeCAD's code and really try to understand it. I guess I have essentially reimplemented my own "clipper" library, but using a more discrete method involving bitmaps and shader operations on the GPU instead to represent the model and stock. I didn't even know how to display a 3D model using opengl when I first began this project, let alone do all of the projection matrix manipulation I do to render the slices you were thinking about with this: 

> I was planning extend it to 3D by simply slicing the model (in pyhton code using existing FreeCAD API) with a plane parallel to XY in small steps across the Z axis (i.e. tool step down) and run this same 2d adaptive for each slice. I think this would be easiest way to get it to work for 3D - by reusing existing 2.5D code.

The paper I link in my readme introduced me to this sort of methodology, generating roughing passes using traditional 3D rendering techniques. It has the benefit of not falling into the trap of how grabbing 2D slices of a model without any thought to parts of a model that might be obstructing parts of the slice above the given 2D slice. All of the math and stuff with involving the depth buffer in OpenGL handles that automatically. I figured I could save the bitmaps for later to perform rest milling as well.

So yeah, if you were wondering why my prototype is completely standalone, it's because I didn't exactly know how to integrate into a larger project quite yet, but I wanted to learn how to do something like this before I got up the courage to really delve into modifying/extending FreeCAD itself, especially with how... different my process is using openGL for offscreen computation.

Thank you for the email address. I may need it, since I don't know a lot of the math very well with what you are doing with clipper for generating potential paths (yet). Hope you have a good day. 

EDIT: BTW, I bet sliptonic wouldn't mind another person working on path if you ever feel like it lol. There is a gitter/matrix server specifically for the path workbench.
